ddl/placement/bundle.go

```diff
@@ -49,6 +49,224 @@ func NewBundle(id int64) *Bundle {
+	}
+
+	if learnerCount > 0 {
+		LearnerRules, err := NewRules(Learner, learnerCount, learnerConstraints)
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid LearnerConstraints", err)
+		}
+		for _, rule := range LearnerRules {
+			for _, cnst := range CommonConstraints {
+				if err := rule.Constraints.Add(cnst); err != nil {
+					return nil, fmt.Errorf("%w: LearnerConstraints conflicts with Constraints", err)
+				}
+			}
+		}
+		Rules = append(Rules, LearnerRules...)
+	}
+
+	return &Bundle{Rules: Rules}, nil
+}
+
+// NewBundleFromSugarOptions will transform syntax sugar options into the bundle.
+func NewBundleFromSugarOptions(primaryRegion string, regions []string, followers uint64, schedule string) (*Bundle, error) {
+	if followers == 0 {
+		followers = 2
+	}
+	var constraints Constraints
+	var err error
+
+	if len(primaryRegion) == 0 {
+		return nil, fmt.Errorf("%w: empty primary region", ErrInvalidPlacemeOptions)
+	}
+
+	Rules := []*Rule{}
+	switch strings.ToLower(schedule) {
+	case "", "even":
+		constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", primaryRegion)})
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid PrimaryRegion '%s'", err, primaryRegion)
+		}
+		Rules = append(Rules, NewRule(Leader, 1, constraints))
+	case "majority_in_primary":
+		// number of quorum members, without the leader
+		primary := (followers + 1) / 2
+		constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", primaryRegion)})
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid 'PrimaryRegion', '%s'", err, primaryRegion)
+		}
+		Rules = append(Rules, NewRule(Leader, 1, constraints))
+		Rules = append(Rules, NewRule(Follower, primary, constraints))
+		// even split the remaining followers
+		followers = followers - primary
+	default:
+		return nil, fmt.Errorf("%w: unsupported schedule %s", ErrInvalidPlacemeOptions, schedule)
+	}
+
+	if uint64(len(regions)) > followers {
+		return nil, fmt.Errorf("%w: remain %d region to schedule, only %d follower left", ErrInvalidPlacemeOptions, uint64(len(regions)), followers)
+	}
+
+	if len(regions) == 0 {
+		constraints, err := NewConstraints(nil)
+		if err != nil {
+			return nil, err
+		}
+		Rules = append(Rules, NewRule(Follower, followers, constraints))
+	} else {
+		count := followers / uint64(len(regions))
+		rem := followers - count*uint64(len(regions))
+		for i, region := range regions {
+			constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", region)})
+			if err != nil {
+				return nil, fmt.Errorf("%w: invalid region of 'Regions', '%s'", err, region)
+			}
+			if i == 0 && rem > 0 {
+				Rules = append(Rules, NewRule(Follower, count+rem, constraints))
+			} else {
+				Rules = append(Rules, NewRule(Follower, count, constraints))
+			}
+		}
+	}
+
+	return &Bundle{Rules: Rules}, nil
+}
+
+// NewBundleFromOptions will transform options into the bundle.
+func NewBundleFromOptions(options []*ast.PlacementOption) (*Bundle, error) {
```

Comment:
Why not use `placementpolicy.PlacementSettings` ? After binding a placement policy to table, we can not get `[]*ast.PlacementOption` from inforschema unless restoring it. I think `placementpolicy.PlacementSettings` is easier to use.

Reply:
This PR is developed very earlier, at the time of no `PlacementSettings`. Furthermore, we need to support direct placement options, not only the named one. It will be addressed later.

Reply:
Addressed.